### PR TITLE
added USB comm option

### PR DESCRIPTION
**Pull request by trestoh**

### CODE REVIEW

File: Common/FeatureIdentification.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    [Guid("4CCF3543-73EF-49D5-8234-FC2FDF647127")]
+    [InterfaceType(ComInterfaceType.InterfaceIsDual)]
+    public interface IFeatureIdentification
+    {
+        /// <summary>board configuration</summary>
+        BOARD_TYPES boardType { get; }
+
+        string firmwarePartNum { get; }
+        string firmwareDesc { get; }
+        bool isSupported { get; }
```
Comment (mzieg) [inline Common/FeatureIdentification.cs:27]: Not yet sure what this is for

File: Common/FeatureIdentification.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                firmwareDesc = "FX2 InGaAs USB Board";
+                defaultPixels = 512;
+                spectraBlockSize = defaultPixels * 2;
+                hasDefaultTECSetpointDegC = true;
+                defaultTECSetpointDegC = -15;
+            }
+            else if (pid == 0x4000)
+            {
+                boardType = BOARD_TYPES.ARM;
+                firmwarePartNum = "170019";
```
Comment (mzieg) [inline Common/FeatureIdentification.cs:89]: 170112

File: Common/FeatureIdentification.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                firmwareDesc = "ARM USB Board";
+            }
+            else if (vid == 0x2457)
+            {
+
+            }
+            else if (pid == 0x0000)
+            {
+                boardType = BOARD_TYPES.STROKER;
```
Comment (mzieg) [inline Common/FeatureIdentification.cs:98]: I would delete this option. Never going to be supported.

File: Common/UInt40.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace EnlightenMAUI.Common
+{
+    /// <summary>
+    /// 40-bit unsigned value, used for many of the laser functions.
+    /// </summary>
+    [ComVisible(true)]
+    [Guid("5BC91277-C373-44F3-8C84-77964365F627")]
+    [InterfaceType(ComInterfaceType.InterfaceIsDual)]
+    public interface IUInt40
```
Comment (mzieg) [inline Common/UInt40.cs:16]: For the record, SiG FPGAs use UInt32 for all modulation commands; there are no 40-bit parameters.

That said, the "extra byte" was necessarily passed as the first byte of the historically-mandatory 8-byte payload block, and is uniformly zero in all anticipated use-cases, which is the default payload value _anyway,_ so there is probably no consequence of continuing to "pretend" these are 40-bit fields on SiG. Other than confusing readers who treat this as current reference code.

File: EnlightenMAUI.csproj [hunk 5]
```diff
@@ -49,13 +49,13 @@
 
 	<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net8.0-android34.0|AnyCPU'">
 	  <IsPublishable>False</IsPublishable>
-	  <ApplicationDisplayVersion>0.5.1</ApplicationDisplayVersion>
+	  <ApplicationDisplayVersion>0.5.0</ApplicationDisplayVersion>
```
Comment (mzieg) [inline EnlightenMAUI.csproj:52]: I disagree that we should go backwards in version when adding a feature. This should probably be 0.5.2, or maybe even 0.6.0.

File: EnlightenMAUI.csproj [hunk 6]
```diff
@@ -77,6 +77,16 @@
 	</ItemGroup>
 
 	<ItemGroup>
+	  <None Remove="libraries\SiG-785\2mmHDPE.csv" />
```
Comment (mzieg) [inline EnlightenMAUI.csproj:80]: A reminder that EnlightenMAUI is open-source, and we should be careful about rolling matching libraries into the distribution. These particular spectra are non-problematic, but that won't always be the case.

File: Models/USBSpectrometer.cs [hunk 7]
```diff
@@ -0,0 +1,696 @@
...
+        internal bool shuttingDown = false;
+        static FeatureIdentification featureIdentification { get; set; }
+        internal Dictionary<Opcodes, byte> cmd = OpcodeHelper.getInstance().getDict();
+        HashSet<Opcodes> armInvertedRetvals = OpcodeHelper.getInstance().getArmInvertedRetvals();
+
+        public bool isConnected = false;
+
+        public bool isStroker { get; protected set; } = false;
+        public virtual bool isARM => featureIdentification.boardType == BOARD_TYPES.ARM;
+        public bool isSiG => eeprom.model.ToLower().Contains("sig") || eeprom.detectorName.ToLower().Contains("imx");
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:34]: || eeprom.model.ToLower().Contains("xs")

File: Models/USBSpectrometer.cs [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+            if (laserExcitationNM > 0)
+                wavenumbers = Util.wavelengthsToWavenumbers(laserExcitationNM, wavelengths);
+            else
+                wavenumbers = null;
+
+            logger.debug("Spectrometer.initAsync: generating pixel axis");
+            generatePixelAxis();
+
+            // set this early so battery and other BLE calls can progress
+            paired = true;
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:116]: Are we using "paired" to indicate "enumerated"? How does this differ from isConnected?

File: Models/USBSpectrometer.cs [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+            gainDb = 8;
+
+            verticalROIStartLine = eeprom.ROIVertRegionStart[0];
+            verticalROIStopLine = eeprom.ROIVertRegionEnd[0];
+
+            logger.info($"initialized {eeprom.serialNumber} {fullModelName}");
+            logger.info($"  detector: {eeprom.detectorName}");
+            logger.info($"  pixels: {pixels}");
+            logger.info($"  integrationTimeMS: {integrationTimeMS}");
+            logger.info($"  gainDb: {gainDb}");
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:142]: So, we're acting as though all USBSpectrometers have gain (dB), but we also explicitly support Hamamatsu CCD (0x1000) and InGaAs (0x2000) models which certainly don't have gain (dB). Long-term, we should consider subclassing USBSpectrometer into CCDUSBSpectrometer, InGaAsUSBSpectrometer, and CMOSUSBSpectrometer (or add a whole bunch of if-then conditional logic to USBSpectrometer).

I'm not treating this as a release blocker, just pointing out a fairly big "to-do" for the backlog.

File: Models/USBSpectrometer.cs [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+            logger.debug("Spectrometer.reset: start");
+            paired = false;
+
+            // Provide some test defaults so we can play with the chart etc while
+            // disconnected.  These will all be overwritten when we read an EEPROM.
+            pixels = 1952;
+            laserExcitationNM = 785.0f;
+            wavelengths = new double[pixels];
+            for (int i = 0; i < pixels; i++)
+                wavelengths[i] = laserExcitationNM + 15 + i / 10.0;
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:191]: Whut

File: Models/USBSpectrometer.cs [hunk 11]
```diff
@@ -0,0 +1,696 @@
...
+            logger.info("sending spectrum trigger");
+            byte[] buf = null;
+            if (isARM)
+                buf = new byte[8];
+
+            logger.debug("sending SW trigger");
+            await sendCmdAsync(Opcodes.ACQUIRE_SPECTRUM, 0, buf: buf);
+
+            byte[] spectrumBuff = new byte[3904];
+            int okI = await udc.BulkTransferAsync(acc.GetInterface(0).GetEndpoint(0), spectrumBuff, 3904, (int)(integrationTimeMS * 8 + 500));
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:455]: I'd replace 3904 with pixels * 2

File: Models/USBSpectrometer.cs [hunk 12]
```diff
@@ -0,0 +1,696 @@
...
+            {
+                logger.info("successfully read {0} bytes: [ {1} ]", okI, String.Join(' ', spectrumBuff));
+            }
+            else
+            {
+                logger.info("failed to read from USB with code {0}", okI);
+            }
+
+            uint[] subspectrum = new uint[1952];
+            for (int i = 0; i < 1952; i++)
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:467]: pixels

File: Models/USBSpectrometer.cs [hunk 13]
```diff
@@ -0,0 +1,696 @@
...
+                subspectrum[i] = (uint)(spectrumBuff[i * 2] | (spectrumBuff[i * 2 + 1] << 8));  // LSB-MSB
+
+            double[] spec = new double[1952];
+
+            for (int i = 0; i < 1952; i++)
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:472]: etc

File: Models/USBSpectrometer.cs [hunk 14]
```diff
@@ -0,0 +1,696 @@
...
+
+            byte[] buf = new byte[bytesToRead];
+
+            bool expectedSuccessResult = true;
+            if (isARM && armInvertedRetvals.Contains(opcode))
+                expectedSuccessResult = !expectedSuccessResult;
+
+            bool result = false;
+
+            int okI = await udc.ControlTransferAsync((UsbAddressing)DEVICE_TO_HOST, cmd[Opcodes.SECOND_TIER_COMMAND], cmd[opcode], wIndex, buf, bytesToRead, 100);
```
Comment (mzieg) [inline Models/USBSpectrometer.cs:537]: okI = okInt?

File: Models/USBViewDevice.cs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        /// <remarks>
+        /// This is not an ideal location, because most of this "BLEDevice" Model 
+        /// relates to A SPECIFIC BLEDevice, and not the general application state of
+        /// being paired. However, it is at least encapsulated in a MODEL, which 
+        /// means that all ViewModels can refer to it; this beats the previous 
+        /// approach of having separate flags in ScopeViewModel and 
+        /// BluetoothViewModel, or having ViewModels reference each other etc. Maybe
+        /// someday we'll add Models.Application[Bluetooth]State, but this is fine
+        /// for now. Making static to distinguish from the instance :-)
+        /// </remarks>
```
Comment (mzieg) [inline Models/USBViewDevice.cs:30]: Maybe remove BLE from comments.

File: Models/USBViewDevice.cs [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        string _Name = "";
+
+        // This is the little hex string displayed under the device name in the
+        // BluetoothView allowing you to distinguish between multiple WP-SiG
+        // devices in Bluetooth range.  Bizarrely, the value won't match in
+        // content or format between iOS and Android :-(
+        //
+        // On iOS, this gives a 128-bit device UUID, e.g. fd7b9fca-3615-da68-03f2-6557e29e2be4
+        // On Android, this gives a less-impressive 48-bit device UUID, e.g. f1e9a7ce0ac8
```
Comment (mzieg) [inline Models/USBViewDevice.cs:72]: delete comment block

File: Pages/BluetoothPage.xaml [hunk 17]
```diff
@@ -8,19 +8,28 @@
...
 
-            <!-- Button Bar -->
-            <Frame CornerRadius="10" BackgroundColor="#666" Padding="2">
-                <Frame CornerRadius="10" BackgroundColor="#444" Padding="2">
-                    <StackLayout Orientation="Horizontal" Padding="2" HorizontalOptions="FillAndExpand" x:DataType="local:ViewModels.BluetoothViewModel">
+            <StackLayout BackgroundColor="#333" HorizontalOptions="Center" Orientation="Horizontal">
+                <Label Text="Bluetooth" VerticalOptions="Center" TextColor="#cccccc"/>
+                <RadioButton IsChecked="{Binding usingBluetooth}" BorderColor="#cccccc"/>
+                <Label Text="USB" VerticalOptions="Center" TextColor="#cccccc"/>
+                <RadioButton IsChecked="{Binding usingUSB}" BorderColor="#cccccc"/>
```
Comment (mzieg) [inline Pages/BluetoothPage.xaml:22]: Interesting. So the user can choose whether the "BluetoothPage" (todo, "ConnectionPage") displays either USB or BLE devices. I like it.

File: ViewModels/BluetoothViewModel.cs [hunk 18]
```diff
@@ -9,21 +9,66 @@
...
+//using Android.Hardware.Usb;
+//using Android.Content;
+using LibUsbDotNet.Main;
+using Android.Content;
+using Android.Hardware.Usb;
+using Android.App;
+using Android.Nfc;
+using Microsoft.Maui.Controls.PlatformConfiguration;
+using Microsoft.Maui;
+using Xamarin.Google.Crypto.Tink.Subtle;
```
Comment (mzieg) [inline ViewModels/BluetoothViewModel.cs:21]: That is the funniest package name I've ever read.

File: ViewModels/BluetoothViewModel.cs [hunk 19]
```diff
@@ -9,21 +9,66 @@
...
+    Android.Hardware.Usb.UsbDevice acc;
 
     public Command scanCmd { get; }
+    public Command scanUSBCmd { get; }
     public Command connectCmd { get; }
+    public Command connectUSBCmd { get; }
 
     IBluetoothLE ble;
     IAdapter adapter;
+    private static string ACTION_USB_PERMISSION = "com.android.example.USB_PERMISSION";
```
Comment (mzieg) [inline ViewModels/BluetoothViewModel.cs:42]: spiffy

File: ViewModels/BluetoothViewModel.cs [hunk 20]
```diff
@@ -290,6 +399,108 @@ private async Task<bool> doScanAsync()
...
+        }
+        catch (Exception ex)
+        {
+            logger.info("USB grab failed with error {0}", ex.Message);
+        }
+        */
+
+        return true;
+        
+    }
```
Comment (mzieg) [inline ViewModels/BluetoothViewModel.cs:503]: This function packs a lot of punch. Nice!

File: ViewModels/BluetoothViewModel.cs [hunk 21]
```diff
@@ -381,232 +601,316 @@ private async Task<bool> doConnectOrDisconnectAsync()
...
-        BLEDevice.paired = false;
         PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(connectButtonBackgroundColor)));
         PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(buttonConnectText)));
         return true;
     }
 
-    async Task<bool> doConnectAsync()
+    /*
+     *  I personally think this should probably live in a model class since nothing going on here needs to be displayed
+     */
```
Comment (mzieg) [inline ViewModels/BluetoothViewModel.cs:721]: Good call.

**Review by mzieg — APPROVED** [review]
Hi Trent, that's a hell of a feature update. Huge win, done quickly. I think there's room for some tidy and refactoring down the road, but this meets the requirement and I really appreciate how quickly you knocked it out. Well done 👍

File: Common/FeatureIdentification.cs (hunk 2 above)
Comment (trestoh) [reply Common/FeatureIdentification.cs:89]: this is a copy paste job, but I'll change it

File: Common/FeatureIdentification.cs (hunk 3 above)
Comment (trestoh) [reply Common/FeatureIdentification.cs:98]: as noted above, but will do

File: Common/UInt40.cs (hunk 4 above)
Comment (trestoh) [reply Common/UInt40.cs:16]: I thought this class was a dependency for another one, but it ended up being for a feature we aren't currently supporting. I'll delete for now and bring back over if we need it

File: EnlightenMAUI.csproj (hunk 5 above)
Comment (trestoh) [reply EnlightenMAUI.csproj:52]: overlooked this one

File: EnlightenMAUI.csproj (hunk 6 above)
Comment (trestoh) [reply EnlightenMAUI.csproj:80]: this is a good note, I will make names opaque when needed (or maybe even completely anonymized) and probably tend towards not storing spectra in the repo

File: Models/USBSpectrometer.cs (hunk 8 above)
Comment (trestoh) [reply Models/USBSpectrometer.cs:116]: I added paired to the base class but not isConnected 😆, yeah these are redundant

File: Models/USBSpectrometer.cs (hunk 9 above)
Comment (trestoh) [reply Models/USBSpectrometer.cs:142]: gain uses the same API for all of those models over USB, does it not? I think this level of abstraction is fine whether we're referring to gainDB or whatever unit CCD/InGaAs use for their gain instead. I don't think we'd gain (haha) much in that class split, though you have a better sense of the differences than I do

File: Models/USBSpectrometer.cs (hunk 10 above)
Comment (trestoh) [reply Models/USBSpectrometer.cs:191]: I copied this from your code! I think it's just so the chart doesn't vomit if there's a non calibrated spec connected

File: Models/USBSpectrometer.cs (hunk 14 above)
Comment (trestoh) [reply Models/USBSpectrometer.cs:537]: yes

File: Models/USBViewDevice.cs (hunk 15 above)
Comment (trestoh) [reply Models/USBViewDevice.cs:30]: done

File: Models/USBViewDevice.cs (hunk 16 above)
Comment (trestoh) [reply Models/USBViewDevice.cs:72]: done

File: Pages/BluetoothPage.xaml (hunk 17 above)
Comment (trestoh) [reply Pages/BluetoothPage.xaml:22]: yes! I don't love how the control looks here but it does the job for now, and I as I said in the commit it isn't tested in USB yet but I'll test when I have the appropriate hardware (next week, I only have a usb micro android right now and no micro to c cable, obviously)

File: ViewModels/BluetoothViewModel.cs (hunk 18 above)
Comment (trestoh) [reply ViewModels/BluetoothViewModel.cs:21]: it is unfortunately unneeded, but I agree

File: ViewModels/BluetoothViewModel.cs (hunk 20 above)
Comment (trestoh) [reply ViewModels/BluetoothViewModel.cs:503]: It's where the magic happens for sure

**Pull request merged**